[Save Data] Move unlocked natures and IVs from DexData to StarterData

MokaStitcher opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-       [Bug]: If the PR is primarily a bug fix
-      [Move]: If a move has new or changed functionality
-   [Ability]: If an ability has new or changed functionality
-      [Item]: For new or modified items
-   [Mystery]: For new or modified Mystery Encounters
-      [Test]: If the PR is primarily adding or modifying tests
-     [UI/UX]: If the PR is changing UI/UX elements
-     [Audio]: If the PR is adding or changing music/sfx
-    [Sprite]: If the PR is adding or changing sprites
-   [Balance]: If the PR is related to game balance
- [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-      [Docs]: If the PR is just adding or modifying documentation (such as tsdocs/code comments)
-    [GitHub]: For changes to GitHub workflows/templates/etc
-      [Misc]: If no other category fits the PR
-->
<!--
Make sure that this PR is not overlapping with someone else's work
Please try to keep the PR self-contained (and small)
-->

> [!WARNING]
> This is the first of 2 PRs making changes to the Dex / System Data, meaning partial loss of compatibility with existing saves.
Since the impacted elements (natures and ivs) are of little importance to current development this PR introduces a temporary "migration" of existing data which consists of unlocking all the elements instead of transferring the current unlocks. This means current saves work fine, and the current `everything` save file can still be used to unlock everything
The next PR will remove that migrator and fully force a reset of current system data, as well as provide a new, clean, `everything` save file.

## What are the changes the user will see?

- existing save files: already unlocked starters will have all their natures unlocked and IVs maxed
- fresh files: starters now get all neutral natures unlocked by default, instead of a single random one
- everything else should be the same except less space used for save file storage

<!-- Summarize what are the changes from a user perspective on the application -->

## Why am I making these changes?

Currently the nature and IVs data is stored in `systemData.dexData`, meaning there is a nature and IV entry for each of the 1100+ Pokémon species. However, unlocked natures and IVs are only relevant when it comes to starters
So, by moving this information from `systemData.dexData` to `systemData.starterData` we change those 1100+ entries to 587, one for each of the starters, resulting in roughly half the storage space used for these attributes

<!--
Explain why you decided to introduce these changes
Does it come from an issue or another PR? Please link it
Explain why you believe this can enhance user experience
-->
<!--
If there are existing GitHub issues related to the PR that would be fixed,
you can add "Fixes #[issue number]" (ie: "Fixes #1234") to link an issue to your PR
so that it will automatically be closed when the PR is merged.
-->

## What are the changes from a developer perspective?

- Move nature and ivs from DexData to StarterData
- Move nature and ivs initialization in `game-data` accordingly
- Update IV scanner and container showing information of newly caught or hatched Pokémon accordingly
- Add temporary data migration unlocking all natures and maxxing out all starter IVs on existing save files only 
> [!NOTE]
> This "migration" does **not** remove the old data from `dexData`, it only populates `starterData`, meaning if you go from this branch back to another one, your old data is preserved for now (and also the save data is larger).
- **UPDATE** Removed the 1 offset in the nature attribute. It is not needed and meant that the attribute for all natures being unlocked was 26 bits long rather than 25

<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->

## Screenshots/Videos

<details><summary>Existing save file</summary>

Other branches:
![image](https://github.com/user-attachments/assets/ed154505-28e5-424e-9aaf-8e1349b8b813)
This branch:
![image](https://github.com/user-attachments/assets/c1dedbc2-a2cc-4605-9bab-f77866ffee81)

</details>

<details><summary>New save file</summary>

After: 15 IV, single neutral nature unlocked:
![image](https://github.com/user-attachments/assets/03545522-6407-46c0-b8f3-b063b6f0e475)
After: 15 IV, all neutral natures unlocked:
![image](https://github.com/user-attachments/assets/41ef1418-aeb2-477b-9dde-b3dfe06c97ff)

</details>

<details><summary>StarterData update when catching/hatching mons</summary>

Natures still get shown as new or not when hatching/catching. IVs get compared to existing IVs when using scanner:
![image](https://github.com/user-attachments/assets/1235eee3-8ee2-4c57-b097-fa663e6e6b52)
![image](https://github.com/user-attachments/assets/f45e26d8-ede9-42f7-9880-dffbb8c4c861)

Catching an Elekid and an Electabuzz:
![image](https://github.com/user-attachments/assets/dddd13c7-66d6-4b61-8ced-d39662514133)
![image](https://github.com/user-attachments/assets/9d64baf0-6257-4710-a519-7fb87578a7d7)
Elekid's IVs are the max of IVs across both catches. Elekid can choose from 2 natures.
Electabuzz only has IVs from the Electabuzz catch, and a single nature
![image](https://github.com/user-attachments/assets/4634d0d7-f5d1-4c13-9a43-3a9849da1705)
![image](https://github.com/user-attachments/assets/a70e0228-3127-425d-a5b8-89b772ee6da8)
</details>

<details><summary>Storage</summary>

**Size of exported fresh file**
Before: 159KB - After: 142KB
**Size of exported advanced file**
Before: 187KB - After: 217KB (because of the temporary migration which doesn't delete the elements from dexData)

</details> 

<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->

## How to test the changes?
- Load the game with an existing save file, and go to starter select
  - All starters should have maxxed IVs and all natures unlocked

- Load the game in a private window to start a fresh file
  - All starters should have a single neutral nature unlocked, and IVs of 15
  - Start a run, catch Pokemon and check for IV and natures:
     - that they are shown properly as new/improved or not, depending
     - that they are unlocked properly back in starter selection
  - Refresh the page to force loading the data. The unlocks should be preserved

You might want to do extra testing around Pokémon with baby forms. Catching a Pikachu should unlock nature and update IVs for both Pikachu and Pichu, while catching a Pichu should only unlocks those things for Pichu itself. Another thing to potentially test for is eggs.

Here are some overrides that can be useful:
```
  STARTING_HELD_ITEMS_OVERRIDE: [{name: "IV_SCANNER", count: 2}],
  //ENEMY_SPECIES_OVERRIDE: SpeciesId.ELECTABUZZ,
  EGG_FREE_GACHA_PULLS_OVERRIDE: true,
  EGG_IMMEDIATE_HATCH_OVERRIDE: true,
  FREE_CANDY_UPGRADE_OVERRIDE: true,
  POKEBALL_OVERRIDE: {
    active: true,
    pokeballs: {
      [PokeballType.POKEBALL]: 5,
      [PokeballType.GREAT_BALL]: 0,
      [PokeballType.ULTRA_BALL]: 0,
      [PokeballType.MASTER_BALL]: 50,
    },
  }
```
<!--
How can a reviewer test your changes once they check out on your branch?
Did you make use of the `src/overrides.ts` file?
Did you introduce any automated tests?
Do the reviewers need to do something special in order to test your changes?
-->

## Checklist

- [] ⚠️ If this is a PR for `main` (such as a hotfix), has the game version been updated (`npm run update-version:patch` / `npm run update version:minor`?
- [X] Otherwise: **I'm using `beta` as my base branch**
- [X] There is no overlap with another PR?
- [X] The PR is self-contained and cannot be split into smaller PRs?
- [X] Have I provided a clear explanation of the changes?
- [X] Have I tested the changes manually?
<!-- We have heavily optimized our test suite, so please actually run the tests :) -->
- [X] Are all unit tests still passing? (`npm run test:silent`)
  - [ ] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?
- [X] Have I provided screenshots/videos of the changes (if applicable)?
  - [] Have I made sure that any UI change works for both UI themes (dark and light)?

#### If there are no locale changes:
- [X] Have I made sure **not** to commit any changes to the locale repo on this branch?
<!-- check the `Files Changed` tab on the PR to be sure. 
`public/locales` should not appear here, or it will create needless conflicts for future PRs and could potentially roll back changes already merged to beta. -->

<!-- How to fix it if no: -->
<!-- #### Using the Command Line:
- Go to https://github.com/Despair-Games/poketernity/tree/beta/public and copy the hash of the current locale commit beta is pointing to
- If the hash corresponds to the latest commit in the locale repo:
  - `npm run update-locales:remote`
  - `git add public/locales`
  - make your commit, push etc
- If it's not the latest commit:
  - `git checkout beta`
  - if not up to date: `git pull`. Otherwise: `npm run update-locales:remote`
  - `git checkout {this pr's branch}`
  - `git add public/locales`
  - make your commit, push etc

 /!\ anyone using another tool, feel free to add instructions there /!\

You may have to do this again later and fix conflicts if beta keeps updating the locale repo.
**When fixing conflicts, make sure you prioritize the latest commit between the one on beta and this branch, to avoid any rollbacks.** -->

MokaStitcher committed:
move nature from dex data to starter data

MokaStitcher committed:
move ivs from dex data to starter data

Tempo-anon commented on src/phases/select-starter-phase.ts:
```diff
@@ -74,7 +74,10 @@ export class SelectStarterPhase extends Phase {
       }
 
-      const starterIvs = dexData[speciesId].ivs.slice(0);
+      // Get ivs from the root species in case of an override to a non starter species
```
Technically this is new functionality when overriding to a different species right?

MokaStitcher replied:
Not really? Not that the previous behavior was documented or predictable in any way. And this one is still not predictable, to me it's kind of the same except on average it will result in higher IVs
Before if you did an override to Charizard it would look for Charizard's ivs, and now Charizard doesn't have IVs so it uses Charmander's, which are always the same, or better than what Charizard's would have been

MokaStitcher replied:
Feel free to elaborate on what you'd want done in this regard, update the override to make this behavior clear?

Tempo-anon replied:
Ah okay. What happens if I choose Charmander as a starter but I use a species override as Lapras?

MokaStitcher replied:
It will use lapras' IVs since the species override is applied earlier in that function (the IVs will be 0 if Lapras hasn't been unlocked yet)

Tempo-anon commented on src/system/game-data.ts:
```diff
@@ -374,6 +375,17 @@ export class GameData {
+          if (!starterData.natureAttr || !starterData.ivs) {
+            const unlocked = starterData.abilityAttr !== 0;
+            // As a placeholder migration, unlock all natures
+            starterData.natureAttr = unlocked ? 67108862 : 0;
```
nit: I know this is temporary code but is that 67108862 value worth keeping around as a constant to represent the complete natureAttr? Also since there are 25 natures in game, shouldn't this be 2^25 (33554432)? 2^26 is 67108864

MokaStitcher replied:
For some reason the flags are offset by 1, so 3554432 doesn't work. I don't know if updating that should be part of this PR though, for now I left it as is because I'm scared to mess things up lol
But I don't think we want to keep that as a constant, it was just a very lazy way to do it in this case. I'll see for a cleaner way when creating the proper everything file

edit: nevermind, I removed the +1 everywhere when manipulating the natureattr, I don't see why it would be needed

Tempo-anon commented on src/system/game-data.ts:
```diff
@@ -1459,16 +1452,33 @@ export class GameData {
   private initStarterData(): void {
     const starterData: StarterData = {};
 
+    // Pick a random neutral nature for each of the default starters
```
While I like this in terms of flavor and uniqueness, I wonder if the extra complexity is needed rather than setting all the starters to a single nature. Or granting all the initial starter all neutral natures

Tempo-anon commented on src/system/game-data.ts:
```diff
@@ -1459,16 +1452,33 @@ export class GameData {
-        abilityAttr: defaultStarterSpecies.includes(speciesId) ? AbilityAttr.ABILITY_1 : 0,
+        abilityAttr: isDefaultStarter ? AbilityAttr.ABILITY_1 : 0,
         passiveAttr: 0,
+        natureAttr: isDefaultStarter ? 1 << (defaultStarterNatures.shift()! + 1) : 0,
```
nit: maybe a comment would be good because `1 << (defaultStarterNatures.shift()! + 1)` is not clear to me at a glance

Tempo-anon reviewed (dismissed):
Looks fine to me for the most part. Just some nits

DayKev committed:
Use `.toContain(x)` instead of `.includes(x)).toBeTruthy()`

`.toContain` will output both the object being checked and the value
being checked against if the `expect` fails

DayKev commented on src/system/game-data.ts:
```diff
@@ -1459,16 +1452,33 @@ export class GameData {
+        }
+      },
+      0,
+      "default",
```
Doesn't using `"default"` as a seed override mean every player will have the same "randomized" set of natures for their initial starters? If we're going to keep the "get a random neutral nature" for the starters I think it shouldn't use `executeWithSeedOffset`, or we should just unlock all neutral natures for the initial starters.

MokaStitcher replied:
No clue, that code wasn't changed, just moved

MokaStitcher replied:
got removed and replaced with unlocking all neutral natures

DayKev commented:
Updated the test (see commit message).

DayKev commented:
Tested catching a Raichu with IVs set to `0, 5, 10, 15, 20, 25` and both Pikachu and Pichu had the correct IVs (and the same nature) in starter select. Reversing the IVs and catching another one also updated the IVs for both properly. 👍

MokaStitcher committed:
remove the 1 offset in natureAttr

MokaStitcher committed:
unlock all neutral natures for fresh file starters

MokaStitcher committed:
Merge branch 'beta' into starter-data-update

flx-sta commented on src/constants/game-constants.ts:
```diff
@@ -24,6 +24,11 @@ export const IV_MIN = 0;
+/**
+ * Value used for the IVs of starters unlocked by default on a fresh file.
+ */
+export const IV_DEFAULT = 15;
```
I think this constants name should be more specialized that it's only for `starters`.
Should also be the actual array I'd say. Maybe we want to alter specific IV aspects in the future and not the whole array.

```suggestion
export const IV_DEFAULT_UNLOCKED_STARTERS = [15, 15, 15, 15, 15, 15];
```

DayKev replied:
I think I'd go with `DEFAULT_STARTER_IVS`.

MokaStitcher replied:
Agree about the name, I was just trying to match the existing constants. Will update tomorrow
Unsure about storing the whole array though.
After having looked at the heap so much I think we need to be a bit more mindful about the amount of data we keep in memory at all times. Sure that single array won't make much of a difference but I don't know about keeping an array that is only ever relevant the very first time the game is ran for a potential future use that is very unlikely to happen

MokaStitcher committed:
Merge branch 'beta' into starter-data-update

MokaStitcher committed:
rename default iv constant

flx-sta commented on src/system/game-data.ts:
```diff
@@ -371,6 +372,18 @@ export class GameData {
           }
         }
 
+        // TODO: Temporary starter data migration, to be removed later
+        const allNaturesAttr = Math.pow(2, getEnumLength(Nature)) - 1;
+        for (const starterData of Object.values(systemData.starterData)) {
+          if (!starterData.natureAttr || !starterData.ivs) {
+            const unlocked = starterData.abilityAttr !== 0;
+            // As a placeholder migration, unlock all natures
+            starterData.natureAttr = unlocked ? allNaturesAttr : 0;
+            // As a placeholder migration, max out all ivs
+            starterData.ivs = Array(6).fill(unlocked ? IV_MAX : IV_MIN);
+          }
+        }
```
Shouldn't this be part of the migration tooling?

MokaStitcher replied:
Not worth the trouble, this is extremely temporary, will be removed in my next PR (more details in the PR description itself)

flx-sta commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -3333,9 +3335,10 @@ export class StarterSelectUiHandler extends MessageUiHandler {
 
-      const caughtAttr = globalScene.gameData.dexData[species.speciesId]?.caughtAttr || BigInt(0);
+      const abilityAttr = starterEntry.abilityAttr;
+      const caughtAttr = dexEntry?.caughtAttr || BigInt(0);
```
Always prefer `??` over `||` unless there is a specific reason to use `||`

```suggestion
      const caughtAttr = dexEntry?.caughtAttr ?? BigInt(0);
```

flx-sta commented on test/system/dex-data.test.ts:
```diff
@@ -44,13 +45,17 @@ describe("Dex Data", () => {
+      // Starters get all neutral natures unlocked
+      const unlockedNatures = gameData.getNaturesForAttr(starterData.natureAttr);
+      expect(unlockedNatures.length).toBe(neutralNatures.length);
+      for (const starterNature of unlockedNatures) {
+        expect(neutralNatures).toContain(starterNature);
+      }
```
If you want to avoid the `for of` loop here use `.every()`

```suggestion
      expect(unlockedNatures.every(nature => neutralNatures.includes(nature))).toBe(true);
```
<sub> I personally find expects in loops odd </sub>

DayKev replied:
We should be careful about how we format the `expect`s, they should be as informative as possible if they fail. If this fails it'll just show "expected `false` to be `true`".

DayKev replied:
It's better to change this to
```ts
expect(unlockedNatures).toEqual(neutralNatures);
```

DayKev replied:
![image](https://github.com/user-attachments/assets/8a46af8f-6bf7-4ff8-bbb6-b49a8e464580)
vs
![image](https://github.com/user-attachments/assets/767aba3c-7ceb-42a2-ad99-b9702b162349)
(Replaced "Quirky with Adamant" in the neutral natures array for demonstration purposes)

flx-sta replied:
Alternatively you can also define your own failure messages as the 2nd argument in `expect`

MokaStitcher committed:
Apply suggestions from code review

Co-authored-by: flx-sta <50131232+flx-sta@users.noreply.github.com>

MokaStitcher committed:
Merge branch 'beta' into starter-data-update

MokaStitcher committed:
Merge branch 'beta' into starter-data-update

DayKev committed:
Update `expect` to provide more informative error message on fail

MokaStitcher committed:
Merge branch 'beta' into starter-data-update

MokaStitcher committed:
add test for updating starter ivs
